src/encode/pattern/mod.rs

```diff
@@ -524,6 +328,184 @@ impl<'a> From<Piece<'a>> for Chunk {
+            let old_buf = mem::take(&mut self.buf);
+            let s = String::from_utf8_lossy(&old_buf[..]).into_owned();
+            let len = s.graphemes(true).count();
+            self.strings_and_styles
+                .push(StringOrStyle::String(GraphemeLenString { len, s }));
+        }
+    }
+
+    fn chunk_end(&mut self) -> io::Result<()> {
+        self.push_string();
+        let total_width = self.compute_width();
+        let mut done = false;
+        if let Some(max_width) = self.params.max_width {
+            if total_width > max_width {
+                if self.params.right_truncate {
+                    self.output_right_truncate(max_width)?;
+                } else {
+                    self.output_left_truncate(total_width, max_width)?;
+                }
+                done = true;
+            }
+        }
+        if let Some(min_width) = self.params.min_width {
+            if total_width < min_width {
+                if self.params.align == Alignment::Left {
+                    self.output_everything()?;
+                    self.output_padding(min_width - total_width)?;
+                } else {
+                    self.output_padding(min_width - total_width)?;
+                    self.output_everything()?;
+                }
+                done = true;
+            }
+        }
+        if !done {
+            // between min and max length
+            self.output_everything()?;
+        }
+        Ok(())
+    }
+
+    fn compute_width(&self) -> usize {
+        let mut size = 0;
+        for x in &self.strings_and_styles {
+            if let StringOrStyle::String(g) = x {
+                size += g.len;
+            }
+        }
+        size
+    }
+
+    fn output_left_truncate(&mut self, total_width: usize, max_width: usize) -> io::Result<()> {
+        let mut to_cut = total_width - max_width;
+        for x in &self.strings_and_styles {
+            match x {
+                StringOrStyle::String(g) => {
+                    if to_cut == 0 {
+                        self.w.write_all(g.s.as_bytes())?;
+                    } else if g.len <= to_cut {
+                        to_cut -= g.len;
+                    } else {
+                        let start = Self::boundary_or(g, to_cut, 0);
+                        self.w.write_all(&g.s.as_bytes()[start..])?;
+                        to_cut = 0;
+                    }
+                }
+                StringOrStyle::Style(s) => self.w.set_style(s)?,
+            }
+        }
+        Ok(())
+    }
+
+    fn boundary_or(g: &GraphemeLenString, count: usize, or: usize) -> usize {
+        let mut cursor = GraphemeCursor::new(0, g.s.len(), true);
+        let s = g.s.as_str();
+        let mut start = 0;
+        for _i in 0..count {
+            let r = cursor.next_boundary(s, 0);
+            if let Ok(Some(x)) = r {
+                start = x;
+            } else {
+                // this should never happen, as we sanitize with to_utf8_lossy
```

Comment:
Are we sure about this? Seems like it could, given any input of `g` and `count` if we rely on assumption that should be at least documented if not checked.

Reply:
We don't rely on that assumption. Clarified the comment.